### PR TITLE
CMake: include third-party library directories by linking to the corresponding library targets

### LINKED ISSUE #2032: CMake: third-party includes should not be set in src/CMakeLists.txt
According to my (incomplete) understanding of CMake, this is not the correct way to do things:

https://github.com/hyrise/hyrise/blob/64ed84e275501bfc6023657ad50cba9fc4cbd29f/src/CMakeLists.txt#L66-L70

Instead, the libraries should be defined as such and added as a dependency.

### PR DESCRIPTION
**Pull request by mweisgut**
This PR improves the current `cmake` files. It fixes #2032, i.e., it removes `include_directories` from the `src/CMakeLists.txt` and adds the corresponding libraries with their include directories instead.

### CODE REVIEW

File: src/test/CMakeLists.txt
Comment (mweisgut) [inline src/test/CMakeLists.txt:276]: can probably be removed

File: src/test/CMakeLists.txt
Comment (mweisgut) [inline src/test/CMakeLists.txt:341]: remove debug prints

File: src/lib/CMakeLists.txt
Comment (mweisgut) [inline src/lib/CMakeLists.txt:725]: should be added as library target

File: src/plugins/CMakeLists.txt
Comment (mweisgut) [inline src/plugins/CMakeLists.txt:16]: Can be private. Setting it to public did not solve the issue.

File: src/CMakeLists.txt
Comment (mweisgut) [inline src/CMakeLists.txt:70]: Should be added by linking to a defined interface library target.

**Comment by mweisgut** [comment]
Closed for now. The failure of clangDebugCoverage needs to be debugged.

**Review by Bouncner — APPROVED** [review]
Looks good to me, but again: I am not very familiar with CMake. I would recommend to wait for a second approve before merging.

File: src/lib/storage/index/b_tree/b_tree_index_impl.hpp
Comment (Bouncner) [inline src/lib/storage/index/b_tree/b_tree_index_impl.hpp:3]: How did you catch that?

File: src/lib/storage/lz4_segment/lz4_encoder.hpp
Comment (Bouncner) [inline src/lib/storage/lz4_segment/lz4_encoder.hpp:4]: I needed to do the same recently as I have updated LZ4. I am wondering why you are doing the same without an update on the LZ4 commit?!?

File: src/lib/storage/index/b_tree/b_tree_index_impl.hpp
Comment (mweisgut) [reply src/lib/storage/index/b_tree/b_tree_index_impl.hpp:3]: Before this PR, we explicitly ignored some clang compiler warnings with `#pragma clang diagnostic ignored`. Now, we add `cpp-btree` as system library to suppress these warnings (see `src/lib/CMakeLists.txt` line 730) as we also did it for other third-party libraries before.

File: src/lib/storage/lz4_segment/lz4_encoder.hpp
Comment (mweisgut) [reply src/lib/storage/lz4_segment/lz4_encoder.hpp:4]: This file is actually of the `zstd` library rather than the `lz4` library.

Before this PR, we manually included `${PROJECT_SOURCE_DIR}/third_party/zstd`, i.e., only the library's root directory. Now, we include `${ZSTD_LIBRARY_DIR}/dictBuilder` (see `third_party/CMakeLists.txt` line 163) as `PUBLIC`.
Even though this line 163 was already present in the code before, the included directories were `PRIVATE`, i.e., they were only included for the source files of the `zstd` library; they were not populated to `hyrise_impl`. Thus, to include the file `zdict.h`, the manually included directory `${PROJECT_SOURCE_DIR}/third_party/zstd` was used. Now, changing the included directories of the `zstd` library to `PUBLIC`, they are also populated to the source files of `hyrise_impl`.

**Comment by Bensk1** [comment]
Tried to build on my Mac (cmake version 3.22.1) and got the following error:

```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
.../hyrise/src/lib/lib_include_dirs
   used as include directory in directory .../hyrise/src/lib
.../hyrise/src/plugins/lib_include_dirs
   used as include directory in directory .../hyrise/src/plugins
   used as include directory in directory .../hyrise/src/plugins
   used as include directory in directory .../hyrise/src/plugins
```

I wonder why the CI didn't catch it. Any reasons for not updating CMake (and clang (one of them runs clang 9)) on our Mac CI hosts? Also asking @Bouncner.

**Comment by mweisgut** [comment]
> Tried to build on my Mac (cmake version 3.22.1) and got the following error for `CXXFLAGS=-fcolor-diagnostics cmake -DCMAKE_C_COMPILER=/usr/local/opt/llvm@12/bin/clang -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm@12/bin/clang++ -GNinja -DCMAKE_BUILD_TYPE=Debug ..`:
> 
> ```
> CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
> Please set them or make sure they are set and tested correctly in the CMake files:
> .../hyrise/src/lib/lib_include_dirs
>    used as include directory in directory .../hyrise/src/lib
> .../hyrise/src/plugins/lib_include_dirs
>    used as include directory in directory .../hyrise/src/plugins
>    used as include directory in directory .../hyrise/src/plugins
>    used as include directory in directory .../hyrise/src/plugins
> ```
> 

I'm trying to reproduce the failure: I upgraded to cmake version 3.22.1 and called your listed cmake command, which succeeded. Building `hyriseTest` with `ninja hyriseTest` also worked fine.

@Bensk1 Do you also use `llvm@12: stable 12.0.1 (bottled)`? What macOS version do you use?

> I wonder why the CI didn't catch it. Any reasons for not updating CMake (and clang (one of them runs clang 9)) on our Mac CI hosts? Also asking @Bouncner.

I see no reason not to upgrade it.

**Comment by Bensk1** [comment]
> @Bensk1 Do you also use `llvm@12: stable 12.0.1 (bottled)`? What macOS version do you use?

Yes, I use the same llvm version. My macOS version is 12.1.

**Comment by mweisgut** [comment]
Closing the PR for now until I revise it.

**Comment by Bensk1** [comment]
After I have resolved a local libTBB issue the generator error is gone (which does not make much sense to me since they seem unrelated).

File: src/CMakeLists.txt
Comment (Bensk1) [inline src/CMakeLists.txt:85]: Can we remove `tpcds-result-reproduction` without any replacement because we don't use it for building anything but just use the result table files from it?

File: third_party/CMakeLists.txt
Comment (Bensk1) [inline third_party/CMakeLists.txt:11]: Should we maybe add one, two sentences of highlevel documentation that there are libraries with built-in CMake support and others that we build manually that will be defined below?

**Review by Bensk1 — APPROVED** [review]
Just two minor questions. Apart from that, I have tested different scenarios on my Mac and everything works fine now!

File: third_party/CMakeLists.txt
Comment (mweisgut) [reply third_party/CMakeLists.txt]: done

File: src/CMakeLists.txt
Comment (mweisgut) [reply src/CMakeLists.txt:85]: Removing the include directory here does not remove the git submodule.
```
[submodule "resources/benchmark/tpcds/tpcds-result-reproduction"]
	path = resources/benchmark/tpcds/tpcds-result-reproduction
	url = https://github.com/hyrise-mp/tpcds-result-reproduction.git
```
Thus, after updating the submodules, the result files are available and can be used for query verification.

Comment (Bensk1) [reply src/CMakeLists.txt:85]: > Removing the include directory here does not remove the git submodule.

Of course not. So the answer to above's question is _yes_.

**Comment by Bensk1** [comment]
Ready for merge.

**Pull request merged**
